packages/react-core/src/Authenticator/hooks/useAuthenticatorRoute/utils.ts

```diff
@@ -0,0 +1,214 @@
+import { AuthenticatorRoute } from '@aws-amplify/ui';
+
+import { RenderNothing } from '../../../components';
+import {
+  AuthenticatorMachineContext,
+  AuthenticatorMachineContextKey,
+  AuthenticatorRouteComponentKey,
+  AuthenticatorRouteComponentName,
+} from '../types';
+
+import {
+  UseAuthenticator,
+  UseAuthenticatorSelector,
+} from '../useAuthenticator';
+import { isComponentRouteKey } from '../utils';
+import { DEFAULT_TOTP_ISSUER, MACHINE_PROP_KEYS } from './constants';
+import { UseAuthenticatorRoute } from './types';
+
+// selects nothing
+const defaultSelector = () => [];
+
+const createSelector =
+  (selectorKeys: AuthenticatorMachineContextKey[]): UseAuthenticatorSelector =>
+  (context) =>
+    selectorKeys.map((key) => context[key]);
+
+export const getRouteSelector = (
+  route: AuthenticatorRoute
+): UseAuthenticatorSelector =>
+  isComponentRouteKey(route)
+    ? createSelector(MACHINE_PROP_KEYS[route])
+    : defaultSelector;
+
+const getSelectedMachineProps = (
+  route: AuthenticatorRouteComponentKey,
+  context: AuthenticatorMachineContext
+) =>
+  MACHINE_PROP_KEYS[route].reduce(
+    (acc, key) => ({ ...acc, [key]: context[key] }),
+    {} as AuthenticatorMachineContext
+  );
+
+export function resolveConfirmResetPasswordRoute<PlatformProps = {}>(
+  Component: UseAuthenticatorRoute<
+    PlatformProps,
+    'ConfirmResetPassword'
+  >['Component'],
+  { fields, ...props }: UseAuthenticator
+): UseAuthenticatorRoute<PlatformProps, 'ConfirmResetPassword'> {
+  return {
+    Component,
+    props: {
+      ...Component,
+      ...getSelectedMachineProps('confirmResetPassword', props),
+      fields,
+    },
+  };
+}
+
+export function resolveConfirmSignInRoute<PlatformProps = {}>(
+  Component: UseAuthenticatorRoute<PlatformProps, 'ConfirmSignIn'>['Component'],
+  { fields, ...props }: UseAuthenticator
+): UseAuthenticatorRoute<PlatformProps, 'ConfirmSignIn'> {
+  const { user, ...machineProps } = getSelectedMachineProps(
+    'confirmSignIn',
+    props
+  );
+
+  return {
+    Component,
+    props: {
+      ...Component,
+      ...machineProps,
+      challengeName: user.challengeName!,
```

Comment:
What is `user.challengeName!`?

Reply:
`challengeName` is optional on `AmplifyUser` since it is not immediately populated.  The `!` let's TS that in this case, the `challengeName` will be populated